opwen_email_server/mailers/shared.py

```diff
@@ -0,0 +1,47 @@
+from typing import Callable
+from typing import Dict
+
+from libcloud.storage.types import ObjectDoesNotExistError
+
+from opwen_email_server.actions import _Action
```

Comment:
Given that we're not using the `_Action` class outside of the `actions` module, let's rename it to `Action` to make it explicit that we allow re-using this class outside of the module and don't treat it like a private implementation detail anymore.

Alternatively (and the more I think about it the better the idea seems to me), we can also move the `ProcessServiceEmail` to `opwen_email_server.actions`... as this code evolved, the class is now nicely abstracted so that it doesn't have any mailer specific logic in it so I don't have any concerns in moving it to the core package. This will then also allow us to fold `opwen_email_server/mailers/connexion.py` into `opwen_email_server/integration/connexion.py` for increased simplicity (especially since we already have the celery wrapper for this class in the `opwen_email_server/integration/celery.py` module anyways). Last but not least we can then move `opwen_email_server/mailers/echo/format.py` to `opwen_email_server/mailers/echo.py` and `opwen_email_server/mailers/registry.py` into `opwen_email_server/mailers/__init__.py` which will give us a nice separation: all the code related to integrating the mailers into the system is found in the usual places where other things hook into the system and all the logic that is specific to any particular mailer is found in the `opwen_email_server/mailers` directory.

I know there has been a bit of back and forth about this, but I believe the exercise was well worth it: we now have the mailers code and hook into the existing codebase in a very nice state (abstractions in one place and concretions in another) which made this nice structure emerge over time. Thoughts?

Reply:
I was actually thinking the same thing about possibly getting rid of shared.py and moving ProcessServiceEmail into actions. Having the dynamic registry lookup made things much easier and was a great idea! I totally agree with this and whoever implements a new service in the future won't even have to look at the celery, connexion and ProcessEmailService stuff.  The structure we have and how the mailers code is tied in (or will be tied in on the next PR 🙂) is very nice and the back and forth nature very much so led to this. It was definitely interesting to go through this process and am happy we did!